### PR TITLE
tools: adding slackbot for PR notifications

### PR DESCRIPTION
**Pull request by alyssawilk**
We've been having PRs slipping through the cracks because repokitten assigns API reviewers and maintainer-on-call sees PRs as assigned and doesn't necessarily tag a maintainer for review.

Also adding hopefully helpful reminders for PRs which get stale.  These will be less useful if folks don't use the waiting tag when waiting for response so will kick off a discussion thread on #maintainers about that

Risk Level: n/a
Testing: manual
Docs Changes: n/a
Release Notes: n/a
co-authored-by: mattklein123

### CODE REVIEW

File: tools/slack_tools/collect_prs.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+import datetime
+import os
+import sys
+
+import github
+
+try:
+    input = raw_input  # Python 2
+except NameError:
```
Comment (phlax) [inline tools/slack_tools/collect_prs.py:14]: this shouldnt be necessary - just assume python3 - or depending on how its triggered use a python3 shebang line

File: tools/slack_tools/collect_prs.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    pass  # Python 3
+
+maintainers = {
```
Comment (phlax) [inline tools/slack_tools/collect_prs.py:17]: as i think this is a constant best to capitalize (for bonus points use a tuple as its immutable and consume to dict later)

File: tools/slack_tools/collect_prs.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        delta_days = delta.days
+        delta_hours = delta.seconds // 3600
+        if (delta < datetime.timedelta(hours=8)):
+            recent_prs.append("PR %s' was updated %s hours ago" % (pr_title, delta_hours))
+            continue
+
+        # If we get to this point, the review may be in SLO - nudge if it's in
+        # SLO, nudge in bold if not.
+        message = ""
+        if (delta < datetime.timedelta(hours=get_slo_hours())):
```
Comment (phlax) [inline tools/slack_tools/collect_prs.py:88]: unnecessary parentheses

File: tools/slack_tools/collect_prs.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+        # If the PR hasn't been updated recently, don't nag.
+        # We can revisit this time, and always warn.
+        now = datetime.datetime.now()
+        time_zone_shift = datetime.timedelta(hours=4)
+        pr_age = pr_info.updated_at - time_zone_shift
+        delta = now - pr_age
+        delta_days = delta.days
+        delta_hours = delta.seconds // 3600
+        if (delta < datetime.timedelta(hours=8)):
```
Comment (phlax) [inline tools/slack_tools/collect_prs.py:81]: unnecessary parentheses

File: tools/slack_tools/collect_prs.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+            if label.name == 'waiting':
+                waiting_prs.append("PR %s' is tagged as waiting" % (pr_title))
+                waiting = True
+                continue
+        if (waiting):
+            continue
+
+        # If the PR hasn't been updated recently, don't nag.
+        # We can revisit this time, and always warn.
+        now = datetime.datetime.now()
```
Comment (alyssawilk) [inline tools/slack_tools/collect_prs.py:75]: If folks are likely to use this as their daily reminder, maybe we should warn for anything assigned?  it is nice to have one place to look through

File: tools/slack_tools/collect_prs.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        pr_url = pr_info.html_url
+
+        # If the PR is waiting, collect it in that list and continue.
+        waiting = False
+        for label in pr_info.labels:
+            if label.name == 'waiting':
+                waiting_prs.append("PR %s' is tagged as waiting" % (pr_title))
+                waiting = True
+                continue
+        if (waiting):
```
Comment (phlax) [inline tools/slack_tools/collect_prs.py:70]: unnecessary parentheses

**Review by alyssawilk — COMMENTED** [review]
huh, I thought I ran format_pre locally before I sent this out.  looking

File: tools/slack_tools/collect_prs.py (hunk 3 above)
Comment (phlax) [reply tools/slack_tools/collect_prs.py:88]: ...also other conditionals below...

File: tools/slack_tools/collect_prs.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+    recent_prs = []
+    # A dict of maintainer : outstanding_pr_string to be sent to slack
+    maintainers_and_prs = {}
+    # A placeholder for unassigned PRs, to be sent to #maintainers eventually
+    maintainers_and_prs['unassigned'] = ""
+    # Out-SLO PRs to be sent to #envoy-maintainer-oncall
+    stalled_prs = ""
+
+    # Snag all PRs, including drafts
+    for pr_info in repo.get_pulls("open"):
```
Comment (phlax) [inline tools/slack_tools/collect_prs.py:58]: this loop is quite long and hard to follow - is it possible to break into some functions - it would make it more expressive and allow a load of dedenting

File: tools/slack_tools/collect_prs.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+        # Add a reminder to each maintainer-assigner on the PR.
+        has_maintainer_assignee = False
+        for assignee_info in pr_info.assignees:
+            assignee = assignee_info.login
+            if assignee not in maintainers:
+                continue
+            has_maintainer_assignee = True
+            if assignee not in maintainers_and_prs.keys():
+                maintainers_and_prs[
+                    assignee] = "Good morning, %s, here are your PR reminders for the day \n" % assignee
```
Comment (phlax) [inline tools/slack_tools/collect_prs.py:110]: nit: whose morning!

File: tools/slack_tools/collect_prs.py [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+        print(line)
+
+    for key in maintainers_and_prs.keys():
+        print(key)
+        print(maintainers_and_prs[key])
+        print('\n\n\n')
+
+
+from slack_sdk import WebClient
+from slack_sdk.errors import SlackApiError
```
Comment (phlax) [inline tools/slack_tools/collect_prs.py:139]: these need to be at the top - but i guess ci has told you that already

File: tools/slack_tools/collect_prs.py [hunk 10]
```diff
@@ -0,0 +1,192 @@
+# Script for collecting PRs in need of review, and informing maintainers via
```
Comment (mattklein123) [inline tools/slack_tools/collect_prs.py:1]: It's up to you, but my preference is to just run this as a daily GH action cron. It looks like it should be pretty easy to use an action with code that exists in the repo: https://docs.github.com/en/actions/learn-github-actions/finding-and-customizing-actions?learn=getting_started#referencing-an-action-in-the-same-repository-where-a-workflow-file-uses-the-action.

Do you want to do this as a follow up or try to get this working now? I can help offline with credentials, etc.

File: tools/slack_tools/collect_prs.sh [hunk 11]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+. tools/shell_utils.sh
```
Comment (phlax) [inline tools/slack_tools/collect_prs.sh:3]: im very much trying to get rid of this - can we use bazel for dependencies?

**Review by phlax — CHANGES_REQUESTED** [review]
hi @alyssawilk ive done a quick first review

its not clear to me how this code is triggered - im happy to review/test further once i understand better

File: tools/slack_tools/collect_prs.sh (hunk 11 above)
Comment (alyssawilk) [reply tools/slack_tools/collect_prs.sh:3]: Do we have examples of scripts that run that way?  my tooling skills are pretty weak so I just copied this from one of the other scrips

File: tools/slack_tools/collect_prs.py (hunk 10 above)
Comment (alyssawilk) [reply tools/slack_tools/collect_prs.py:1]: I think follow-up if that's OK

**Review by alyssawilk — COMMENTED** [review]
ah, CI failure is dependabot.  I'm going to fix in the next push becasue I'm not sure if the way phlax is suggesting I rework deps will change

File: tools/slack_tools/collect_prs.py [hunk 12]
```diff
@@ -8,13 +8,10 @@
...
+from slack_sdk import WebClient
+from slack_sdk.errors import SlackApiError
 
-try:
-    input = raw_input  # Python 2
-except NameError:
-    pass  # Python 3
-
-maintainers = {
+Maintainers = {
```
Comment (phlax) [inline tools/slack_tools/collect_prs.py:14]: sorry to be annoying - the convention for constants in python is  all caps - ie `MAINTAINERS`

File: tools/slack_tools/collect_prs.sh (hunk 11 above)
Comment (phlax) [reply tools/slack_tools/collect_prs.sh:3]: i can follow up after and shift this to bazel if thats helpful

File: tools/slack_tools/collect_prs.py (hunk 12 above)
Comment (alyssawilk) [reply tools/slack_tools/collect_prs.py]: haha, not annoying at all.  I thought that request was kinda weird because I read it as Maintainers - makes much more sense w.r.t. python style now I'm not misreading it!

File: tools/slack_tools/collect_prs.py (hunk 5 above)
Comment (alyssawilk) [reply tools/slack_tools/collect_prs.py:72]: Actually I'm just going to update this one now.

Thing I don't understand, how the datetime returned by github doesn't come with a time zone - I did not expect to have to do the shifting manually but the timestamps were off unless I did.  If someone with better python skills has ideas lmk

File: tools/slack_tools/collect_prs.py (hunk 10 above)
Comment (mattklein123) [reply tools/slack_tools/collect_prs.py:1]: Sure that's fine. Per offline convo I will help set up a GH action template that we can move this to in parallel.

File: tools/slack_tools/collect_prs.sh (hunk 11 above)
Comment (mattklein123) [reply tools/slack_tools/collect_prs.sh:3]: Ultimately we are going to run this as a GH action. I think it will be much easier if we just stick with raw pip for this tool since I think it will be trivial to get running in GHA. So I think the end result here is we will lose the SH file and just keep the requirements.txt file. I will follow up once I do a template for GHA.

Comment (phlax) [reply tools/slack_tools/collect_prs.sh:3]: if we are using gh actions to trigger this we probs want to use something like this i think https://github.com/marketplace/actions/pip-installer

**Comment by mattklein123** [comment]
Cross referencing https://github.com/envoyproxy/envoy/pull/16756/files

**Comment by alyssawilk** [comment]
Any other requests for the scripts?

If not should I move them to .github/actions/pr_notifier/ ?

**Comment by mattklein123** [comment]
@alyssawilk I would say let's move it and wire it up. We can launch and iterate. I will help you with the secret stuff offline.

File: .github/workflows/pr_notifier.yml [hunk 13]
```diff
@@ -0,0 +1,22 @@
+on:
```
Comment (mattklein123) [inline .github/workflows/pr_notifier.yml:1]: I would recommend leaving the manual trigger option that I had in the example. This will allow you to manually trigger it from the UI to test it and also if for whatever reason the run fails and we want to run it.

File: .github/workflows/pr_notifier.yml [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+      - name: Set up Python 3.8
+        uses: actions/setup-python@v2
+        with:
+          python-version: '3.8'
+          architecture: 'x64'
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install -r ./.github/actions/pr_notifier/requirements.txt
+      - name: Display Python version
```
Comment (mattklein123) [inline .github/workflows/pr_notifier.yml]: nit: update to better text

Comment (mattklein123) [inline .github/workflows/pr_notifier.yml]: You will need to add the env var for the token secret here.

**Review by mattklein123 — DISMISSED** [review]
Awesome!!! I would just force merge it and you can test it manually.

**Review by phlax — APPROVED** [review]

**Comment by phlax** [comment]
@alyssawilk this still needs the dependabot setup i think

**Review by mattklein123 — DISMISSED** [review]

**Review by mattklein123 — APPROVED** [review]
